crates/dbsp/src/algebra/decimal.rs

```diff
@@ -0,0 +1,607 @@
+            write!(f, "{}", self.0)
+        }
+    }
+
+    impl Default for Scale {
+        fn default() -> Self {
+            Self(MAX_DECIMAL_SCALE)
+        }
+    }
+}
+
+use inner_types::*;
+
+/// A wrapper type around [`rust_decimal::Decimal`].
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, SizeOf)]
+#[cfg_attr(feature = "with-serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "with-serde", serde(transparent))]
+#[derive(rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)]
+pub struct SQLDecimal {
+    inner: Decimal,
+    #[cfg_attr(feature = "with-serde", serde(skip, default))]
+    max_precision: Precision,
+    #[cfg_attr(feature = "with-serde", serde(skip, default))]
+    max_scale: Scale,
+}
+
+impl FromStr for SQLDecimal {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let inner = Decimal::from_str(s).map_err(|e| e.to_string())?;
+
+        Ok(Self {
+            inner,
+            max_precision: Precision::max(),
+            max_scale: Scale::try_from(inner.scale())?,
+        })
+    }
+}
+
+impl TryFrom<Decimal> for SQLDecimal {
+    type Error = String;
+
+    fn try_from(mut inner: Decimal) -> Result<Self, Self::Error> {
+        let scale = Scale::max();
+        scale.apply(&mut inner);
+
+        let precision = Precision::max();
+
+        Ok(Self {
+            inner,
+            max_precision: precision,
+            max_scale: scale,
+        })
+    }
+}
+
+impl SQLDecimal {
+    pub fn new(precision: u32, scale: u32) -> Self {
+        let inner = Default::default();
+
+        let scale = Scale::try_from(scale).unwrap();
+        let precision = Precision::try_from(precision).unwrap();
+
+        Self {
+            inner,
+            max_precision: precision,
+            max_scale: scale,
+        }
+    }
+
+    pub fn from_str_with_precision_scale(
+        s: &str,
+        _precision: u32,
+        scale: u32,
+    ) -> Result<Self, String> {
+        let mut inner = Decimal::from_str(s).map_err(|e| e.to_string())?;
+
+        // TODO: use the precision value from the parameter
+        let max_precision = Precision::max();
+
+        let max_scale = Scale::try_from(scale)?;
+
+        inner.rescale(max_scale.inner());
+
+        Ok(Self {
+            inner,
+            max_precision,
+            max_scale,
+        })
+    }
+
+    pub fn rescale(mut self, val: u32) -> Result<Self, String> {
+        self.max_scale = Scale::try_from(val)?;
+        self.max_scale.apply(&mut self.inner);
+
+        Ok(self)
+    }
+
+    pub fn precision(mut self, val: u32) -> Result<Self, String> {
+        self.max_precision = Precision::try_from(val)?;
+
+        Ok(self)
+    }
+
+    fn apply_scaling_rules(&mut self) {
+        self.max_scale.apply(&mut self.inner)
+    }
+
+    pub fn trunc(&self) -> Self {
+        let mut out = *self;
+        out.inner = out.inner.trunc();
+        out.apply_scaling_rules();
```

Comment:
why is this necessary?

Reply:
The trunc method is necessary as it is used while casting, but yeah, we don't need to apply the scaling rules there.